ex5/y_ogawa/main.py

```diff
@@ -0,0 +1,348 @@
+"""GMMを用いたデータフィッティングを行う."""
+
+import argparse
+
+import matplotlib.pyplot as plt
+import matplotlib.ticker as ticker
+import numpy as np
+from scipy.stats import multivariate_normal
+
+import ex4
+
+
+def parse_args():
+    """引数の取得を行う.
+
+    filename : 読み込むファイル名
+    n_cluster : クラスター数
+    """
+    parser = argparse.ArgumentParser(description="GMMを用いたマッピングを行う")
+    parser.add_argument("--filename", type=str, required=True, help="name of file")
+    parser.add_argument(
+        "--n_components", type=int, required=True, help="number of component"
+    )
+    return parser.parse_args()
+
+
+def random_initialize_para(data: np.ndarray, n_components: int) -> np.ndarray:
+    """ランダムでパラメータを初期化する.
+
+    Args:
+        data (np.ndarray): データ
+        n_components (int): 混合数
+
+    Returns:
+        weights: 混合係数
+        means: 平均
+        covariances: 共分散行列
+    """
+    n_features = data.shape[1]
+
+    # 混合係数をランダムに初期化 (正規化する)
+    weights = np.random.rand(n_components)
+    weights /= weights.sum()
+
+    # 平均値をランダムに初期化
+    means = np.random.rand(n_components, n_features) * (
+        data.max(axis=0) - data.min(axis=0)
+    ) + data.min(axis=0)
+
+    # 共分散行列のランダム初期化（数値安定性のため小さな値を加える）
+    covariances = np.zeros((n_components, n_features, n_features))
+    for i in range(n_components):
+        covariances[i] = np.eye(n_features) * (np.random.rand() + 1e-6)
```

Comment:
```suggestion
    covariances = np.eye(n_features) * (np.random.rand(n_components, 1, 1) + 1e-6)
```
for文を使わず1行にまとめることができます。
np.random.rand(n_components, 1, 1)により、(3, 1, 1)の行列を作成します。
その後、np.eye(n_features) * (np.random.rand() + 1e-6)
によって(2, 2) * (3, 1, 1)の演算になります。
このときnumpyでは、異なるサイズの行列でも演算ができるようブロードキャストという機能が適用され、
各行列は(3, 2, 2)に拡張されます。
よって演算は(3, 2, 2)*(3, 2, 2)となり、返り値も(3, 2, 2)になります